### ISSUE TITLE
Option to skip replacing metadata variables in the body file

### ISSUE THREAD

**Issue by waweber** [jgriff/http-resource#7]
I'm trying to use this resource to upload a built artifact to a repository via HTTP PUT. It works, but there are a couple issues:

1. If the file is binary, the `local tmp="$(cat "$target_file")"` in `replaceBuildMetadataInFile` seems to complain about null bytes (`warning: command substitution: ignored null byte in input`)
2. If the file is particularly large, replacing the metadata is very slow and uses a lot of memory as it reads in the entire file to do the substitution.

These both could be avoided if the build metadata was skipped when the file is specified as binary, or if a setting allowed skipping this.

**Comment by jgriff** [comment]
Makes sense.  Given the potential impact, I think we should also make the build metadata variable substitution an "opt-in" thing (instead of default behavior).  

What if we added a new `build_metadata` option, and also allow param override in each `get` or `put` step.

* `build_metadata`: _Optional._  List of component(s) to perform  [build metadata variable](https://concourse-ci.org/implementing-resource-types.html#resource-metadata) substitution on.  Can be overridden as a param to each `get` or `put` step.  Valid values include:

  * `headers`: Substitute variables used in the `headers`.
  ```yaml
  source:
    build_metadata: [headers]
    headers:
      X-Build-Logs: $ATC_EXTERNAL_URL/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME
  ```
  * `body`:  Substitute variables in `file` and `text` content.
  ```yaml
  source:
    build_metadata: [body]
    text: |
      See build logs at $ATC_EXTERNAL_URL/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME.
  ```
By default, no substitution is performed.

## Examples

```yaml
# no substitution (default)
source:

# substitute headers ONLY
source:
  build_metadata: [headers]

# substitute body ONLY
source:
  build_metadata: [body]

# substitute both
source:
  build_metadata: [headers,body]

# step override - substitute headers (only) for PUT
jobs:
  - plan:
      - put: my-http-resource
        params:
          build_metadata: [headers]
```

**Comment by jgriff** [comment]
This is now available as of `latest` and [`0.5.0`](https://github.com/jgriff/http-resource/releases/tag/v0.5.0).
